media/au.ksy

```diff
@@ -0,0 +1,77 @@
+meta:
+  id: au
+  title: AU audio format
+  file-extension:
+    - au
+    - snd
+  xref:
+    justsolve: AU
+    mime:
+      - audio/basic
+    pronom:
+      - x-fmt/139
+    wikidata:
+      - Q672985
+  license: CC0-1.0
+  ks-version: 0.9
+  endian: be
+doc: |
+  Test files for AU can be found in the Python Git repository:
+
+  https://github.com/python/cpython/tree/b8a7daf077dab18e9e3701c5380b542ae0aa9a94/Lib/test/sndhdrdata
+
+  as well as:
+
+  ftp://ftp-ccrma.stanford.edu/pub/Lisp/sf.tar.gz
+doc-ref:
+  - https://en.wikipedia.org/wiki/Au_file_format
+  - https://web.archive.org/web/20121028010008/http://www-mmsp.ece.mcgill.ca/Documents/AudioFormats/AU/AU.html
+seq:
+  - id: header
+    type: header
+  - id: description
+    type: strz
```

Comment:
It is incorrect, anything can be here.

Reply:
Also, `header_size` means it must be a part of a header, not a field after it.

Reply:
The name "header_size" is directly copied from the specifications (there it is "hdr_size") and it actually is part of the header, so I have no idea what you are talking about.

Reply:
I have already given you the snippet about how IMHO it should look like:

* `header_size` is within the header, but still determines `header` size.
* `sizeof<>` is used, no hardcoded handcoded numbers

Reply:
> It is incorrect, anything can be here.

Although the documentation says this:

```
Further, there are no absolute guarantees that the 'info' is ASCII text, (non-ASCII info may eventually be far more useful anyway).
```

but the SunOS manpage that is being referred to ( audio_filehdr.3 ) specifically mentions a "char *infop" argument to the functions and says:

```
The infop argument, if not NULL, specifies the address of  a
buffer, whose size is given by ilen, into which the informa-
tion field of the audio  file  is  copied.   The  buffer  is
always  returned  NULL-terminated,  even if it was not large
enough to hold the entire file information field.
```

So even though they might have envisioned something else than ASCII this apparently wasn't implemented.

Reply:
> I have already given you the snippet about how IMHO it should look like:
> 
>     * `header_size` is within the header, but still determines `header` size.
> 
>     * `sizeof<>` is used, no hardcoded handcoded numbers
> 
>     * header can be extended by anyone. If you don't know how to parse the extension, it is safely ignored, otherwise you can add code into the end of header parsing the stuff using `if`, but it would require some info passed, either off-band (file extension) or maybe by modifying signature.

I don't think you understand what I did. The only hardcoded value is that the minimum value of header_size should be 24 because that is what directly follows from the specification. Any file that has a value lower than 24 in "header_size" should be rejected. Please read the AU specification.

Reply:
I understand what you did. But`header_size` says that it is a size of a header, not a size of a header with something else after it. So a field of `header` type should occupy that size. You can add something after it, but to have there a `strz` with ASCII you should know ahead of time it is there, something else can be there, the format is forward compatible, it means third-party devs could have implemented own extensions before the spec mentioned ASCII, the format using ASCII text there is only one case of a valid au file.

>So even though they might have envisioned something else than ASCII this apparently wasn't implemented.

>The variable-length annotation field is currently ignored by most audio applications. It is present for interchange compatibility.  In the future, applications may use this field to store structured data annotation, such as markers and edit points.

Reply:
So if I rename it from header_size to "data_offset" it should be fine then :-P

The specifications that I found are actually conflicting. One says it is the header size, the other says that it is the size of the data and that there can be data between the header (24 bytes) and the start of the audio data. The original header file calls it "hdr_size". I'd be more than happy to rename it. Right now we are really splitting hairs.

Reply:
I agree with @KOLANICH. In principle, I'm not a fan of the `valid/min` check that merely emulates the standard EOF checks done automatically by the Kaitai Struct runtime libraries:

https://github.com/kaitai-io/kaitai_struct_formats/blob/98f38afdde472b97aad9f79bf8b3b16e4fda2aa9/media/au.ksy#L38-L41

I think it's better to put the header into a substream with the length of `header_size` bytes. If `header_size < 24`, the parsing will just fail with an `EOFError` pointing to the first element that couldn't fit into the `header_size`-byte substream.

https://github.com/kaitai-io/kaitai_struct_formats/blob/98f38afdde472b97aad9f79bf8b3b16e4fda2aa9/media/au.ksy#L34-L50